### ISSUE TITLE
Very huge numbers written by OMEdit (regression)

### ISSUE THREAD

**Issue by max-privato** [OpenModelica/OpenModelica#9132]
### Description
I remember something similar being discussed months ago, and some activity is going on in the direction of using good representations of numbers
However, I think this case is so surprising, and potentially easy to solve, that maybe it deserves a fast intervention. 

Consider the following model:
```
model Test
  Modelica.Blocks.Math.Gain gain(k = 1e8) annotation(
    Placement(visible = true, transformation(origin = {-10, 12}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
annotation(
    uses(Modelica(version = "4.0.0")));
end Test;
```

The gain value is a very large number, and every engineer immediately understands that it is one hundred million.
Unfortunately, OMEdit renders this as 100000000, which obliges to count several zeros and offers the impression of some sloppiness of OMEdit.
I know that OMEdit has difficulties in using nice formatting of numbers, but I think that at least when the user writing is already in exponential form, it should keep it that way.
**Note that in OM 1.18.0 the number is correctly shown as 3e8.**

### Expected Behavior
Numbers must possibly be concise, especially when this can be attained without loss of precision. 
I know that this, in a general way, requires some cleverness which insofar is not in present OMEdit. But at least when the user uses exponential form explicitly, I think that should be kept.

### Version and OS
1.19.0 Win10 64 bit

**Comment by max-privato** [comment]
I hoped this to be fixed in 1.19.2, but was not.
Is it planned somewhere?

**Comment by casella** [comment]
Sorry, we had to ship 1.19.2 before people started to go on holiday.

I'm not sure why this worked previously, this may be investigated. This is a bit weird, because k is a Real parameter, and 1e8 is a Real literal (not an Integer), so it may be possibly rendered as 100000000.0 (which would also be ugly), but not as 100000000. 

That said, there are many similar issues in OMC, e.g. #7465, #8860, #8865. I believe they should all be resolved systematically using the Ryu library, rather than with ad-hoc patches.

@adeas31 and @perost are now off for a while - when he comes back he will be busy with #8425, which will solve a lot of historical issues. Once they're done with that, they could take care of this.

@max-privato, do you mind reminding us around October?

**Comment by max-privato** [comment]
I've have a personal hatred towards badly written numbers and the need to count zeros to understand the real size of a number.
I think this case is a very simple one: you've to  just fornat a real number in exponential form when appropriate.
I've read the discussion about Ryu library, but that regards cases where rundoffs due to conversion from decimal to binary and back again to decimal form create more difficulties, such as a number originally written as 1.0 rendered as 1.00000001.

This ticket's case is much more trivial and I think could be resolved changing just one or very few rows of code.

On the other hand, I would be ashamed to propose this OM to my students that I train to be careful about how to write numbers
Therefore for next academic year I'll stick to OM 1.18.0, and so, as far as I'm concerned, you'll have plenty of time in playing with your display algorithms.

**Comment by adrpo** [comment]
> Unfortunately, OMEdit renders this as 100000000

We need more information. Where does OMEdit displays it like that? In the text? On the diagram? In the dialogs? On the Plot?

**Comment by adrpo** [comment]
I guess we can find out what changed in OMC and OMEdit to get to this behavior and fix it as that should not be that complicated.

**Comment by adrpo** [comment]
On the diagram is fine:
![image](https://user-images.githubusercontent.com/4192979/179207338-84f47ccf-8f77-4f34-ad8d-2642746c91e7.png)
In the dialog is fine as well:
![image](https://user-images.githubusercontent.com/4192979/179207422-1430518a-8ae5-4144-88f5-aecf604d7773.png)
After adding a space and saving it the diagram changes to:
![image](https://user-images.githubusercontent.com/4192979/179207609-a0c1f9e7-7062-4bd3-ab1d-c7b3b05b951f.png)

So it seems the interaction OMEdit <-> OMC affects that number. I'll do some more hunting to see what's going on.

**Comment by adrpo** [comment]
When restarting OMEdit and loading the model:
[TestBigNumbers.mo.txt](https://github.com/OpenModelica/OpenModelica/files/9120180/TestBigNumbers.mo.txt)
I directly get the bad behavior. The problem seems to be in OMEdit as OMC returns the modifier on `k` as `1e8`:
[omeditcommunication.zip](https://github.com/OpenModelica/OpenModelica/files/9120192/omeditcommunication.zip)

**Comment by casella** [comment]
Thanks @adrpo. This should be easy to fix.

@max-privato if the fix is an easy one, we can make a 1.19.3 release end of August. When are you starting you course?

**Comment by casella** [comment]
@max-privato if you want to use a previous version, I would suggest [1.18.1](https://trac.openmodelica.org/OpenModelica/wiki/ReleaseNotes/1.18.1), which has several important bugfixes over 1.18.0. Unless this thing was broken between 1.18.0 and 1.18.1, in which case it will be very easy to find out why this happened 😅

**Comment by adrpo** [comment]
Ivl even found why it happens. It has to do with unit conversion which is surprising as k has no unit (or actually 1). I will make a PR with the fix soon.

**Comment by adrpo** [comment]
It goes in here because unit (I think) is not empty but is 1:
https://github.com/OpenModelica/OpenModelica/blob/master/OMEdit/OMEditLIB/Annotations/TextAnnotation.cpp#L592
And because it does some calculations it changes the 1e8 to a number and back to text which generates this bad behavior.

**Comment by adrpo** [comment]
Currently, as you can see in the last picture above we display the number and then a space followed by the unit. However in this case we have `100000000 1` which just looks super weird, should we display the unit if is `1`?!

**Comment by adrpo** [comment]
With the PR #9238 we get:
![image](https://user-images.githubusercontent.com/4192979/179320951-a6ed2ef6-2df9-4d82-abab-d52252a74ee9.png)
If unit is `1` we will not display that in the diagram anymore.

**Comment by adrpo** [comment]
Also, if we do conversion we use the previous value to see if there is an `e` in it and if so we display it in the shortest possible way (Qt has a formating option for that: QLocale::FloatingPointShortest).

**Comment by adrpo** [comment]
If there are other places in OMEdit that have this display issue besides the one I fixed now, let us know and we can fix those rather easily.

**Comment by adrpo** [comment]
PR #9239 is applying this fix to the maintenance/v1.19 branch.

**Comment by adrpo** [comment]
When the fix will be in v1.19.3 when we make it.

**Comment by casella** [comment]
> I even found why it happens. It has to do with unit conversion which is surprising as k has no unit

As far as I understand it should have unit="", not unit="1".

**Comment by casella** [comment]
> It goes in here because unit (I think) is not empty but is 1: https://github.com/OpenModelica/OpenModelica/blob/master/OMEdit/OMEditLIB/Annotations/TextAnnotation.cpp#L592 

@adrpo, I checked and indeed k has unit="1", see the [source code](https://github.com/modelica/ModelicaStandardLibrary/blob/8f8ace356bd239776dc778b3369c8600a39aa569/Modelica/Blocks/Math.mo#L543).

I'm not sure what is the rationale behind that, to me this is questionable, but anyway the number formatting shouldn't be changed even in that case.

If there are some computations performed (why?), then Ryu would come back into play, because it would re-format the result in a reasonable way, even though it's not exactly how it was originally input.

**Comment by casella** [comment]
> With the PR #9238 we get: ![image](https://user-images.githubusercontent.com/4192979/179320951-a6ed2ef6-2df9-4d82-abab-d52252a74ee9.png) If unit is `1` we will not display that in the diagram anymore.

Sounds reasonable. One could display p.u. but in some cases that would be weird, so I'm fine with this solution.

**Comment by casella** [comment]
> Also, if we do conversion we use the previous value to see if there is an `e` in it and if so we display it in the shortest possible way (Qt has a formating option for that: QLocale::FloatingPointShortest).

Sounds good as well 😃

**Comment by casella** [comment]
@max-privato, we can easily make a 1.19.3 release with whatever is containd on the maintenance/v1.19 branch, that is no big deal.  

I would suggest that you test 1.19.2 and make sure it fits your teaching needs (besides this glitch). If you find it doesn't because of some other (small) issue, we can try to fix that and include it in 1.19.3. 

Please remind us about releasing 1.19.3 around Sep 10, so we have enough time do do that. By then, there should actually be a 1.20.0 beta1 available, but I guess you want to use a well-tested version, and you don't want to depend on us being timely with releases 😅

**Comment by max-privato** [comment]
> With the PR #9238 we get: ![image](https://user-images.githubusercontent.com/4192979/179320951-a6ed2ef6-2df9-4d82-abab-d52252a74ee9.png) If unit is `1` we will not display that in the diagram anymore.

Ah Sorry, I lost completely the discussion from this point on because when tickets are closed they are not visible on GitHub anymore, by default(for this I miss the old track!)
I'm on vacation now, but when I'm back, I'll make my tests and comment if necessary.
Thank you all.

**Comment by max-privato** [comment]
> @max-privato, we can easily make a 1.19.3 release with whatever is containd on the maintenance/v1.19 branch, that is no big deal.
> 
> I would suggest that you test 1.19.2 and make sure it fits your teaching needs (besides this glitch). If you find it doesn't because of some other (small) issue, we can try to fix that and include it in 1.19.3.
I'll do starting from August, 16.
> 
> Please remind us about releasing 1.19.3 around Sep 10, so we have enough time do do that. By then, there should actually be a 1.20.0 beta1 available, but I guess you want to use a well-tested version, and you don't want to depend on us being timely with releases 😅

In case 1.20.0.beta1 is available in time, I can test all the models I use for teaching with it, and decide whether to prompt you for 1.19.3, or use directly 1.20.0 beta1.
Thank you for taking care of this issue, which I think is minor for many of us, but, as I said, very important for me.

**Comment by max-privato** [comment]
> 
> I would suggest that you test 1.19.2 and make sure it fits your teaching needs (besides this glitch). If you find it doesn't because of some other (small) issue, we can try to fix that and include it in 1.19.3.
> 
To use OM to run Power flow cases of PowerGrids Library I need the following ticket solved.
https://github.com/OpenModelica/OpenModelica/issues/9291

So, in case it is compatible with your schedule, it would be great to have it fixed by rel. 1.19.3.
